### PR TITLE
Add WHERE ternary (or trinary?) op

### LINKED ISSUE #942: Tensor.where with infinity values returns NAN
When using the where operation with a tensor that contains infinity nan is returned.  Seems the issue is in the false condition as float("inf")*0 = NaN 

This behavior can be observed on all runtimes except for METAL. 

GPU=1 python3
```
from tinygrad.tensor import Tensor
x = Tensor.full((3, 3), float("inf"))
(x < 0).where(x, 1).numpy()
array([[nan, nan, nan],
       [nan, nan, nan],
       [nan, nan, nan]], dtype=float32)
```

### PR DESCRIPTION
**Pull request by akretz**
This addresses #1154 and fixes #942. I've only added the op to the tested backends, but adding it to other ones should be pretty straightforward.

Maybe it makes sense to automatically fall back to the current implementation of WHERE in cases where the backend doesn't support that op? I guess such a logic could be added in `ops.py`.

### CODE REVIEW

**Comment by wozeparrot** [comment]
Should update docs like in #1195

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/assembly_ptx.py           2               2
tinygrad/codegen/assembly_rdna.py          3               3
tinygrad/codegen/cstyle.py                 3               2
tinygrad/codegen/linearizer.py             5               5
tinygrad/codegen/llvmir.py                 3               2
tinygrad/graph.py                          3               3
tinygrad/lazy.py                           3               2
tinygrad/mlops.py                         15               1
tinygrad/ops.py                            7               6
tinygrad/runtime/ops_cpu.py                2               2
tinygrad/runtime/ops_torch.py              3               2
tinygrad/tensor.py                        16              22
------------------------------------------------------------
total                                     65              52
------------------------------------------------------------
lines added in the tinygrad folder: 13
```

File: tinygrad/mlops.py [hunk 1]
```diff
@@ -161,6 +161,20 @@ def backward(self, grad_output:LazyBuffer) -> Tuple[Optional[LazyBuffer], Option
     return grad_output.binary_op(BinaryOps.DIV, self.y) if self.needs_input_grad[0] else None, \
            grad_output.const_like(0).binary_op(BinaryOps.SUB, grad_output).binary_op(BinaryOps.MUL, self.x).binary_op(BinaryOps.DIV, self.y.binary_op(BinaryOps.MUL, self.y)) if self.needs_input_grad[1] else None
 
+# ************* ternary ops *************
+
+class Where(Function):
+  __slots__ = "x", "y", "z", "ret"
```
Comment (wozeparrot) [inline tinygrad/mlops.py:167]: we should only need to save `x` here.

**Comment by akretz** [comment]
> Should update docs like in #1195

Right, I just saw @flammit beat me to it by 5 min.

**Comment by wozeparrot** [comment]
Tests failing

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -529,31 +527,27 @@ def mish(self): return self * self.softplus().tanh()
   def softplus(self, beta=1): return (1/beta) * (1 + (self*beta).exp()).log()
   def softsign(self): return self / (1 + self.abs())
 
-  # ***** broadcasted binary mlops *****
+  # ***** broadcasted non-unary mlops *****
 
-  def _broadcasted(self, fxn:Type[Function], other:Union[Tensor, float], reverse:bool=False) -> Tensor:
+  def _broadcasted(self, fxn:Type[Function], *other:Union[Tensor, float], reverse:bool=False) -> Tensor:
```
Comment (geohot) [inline tinygrad/tensor.py:532]: It's slower to make these lists. Is there a simpler refactor?

**Comment by geohot** [comment]
Can you work with #1195 to make one great PR?

**Comment by flammit** [comment]
> Can you work with #1195 to make one great PR?

I've submitted a PR to this branch which merges the changes and keep broadcast as it was.

**Comment by akretz** [comment]
> > Can you work with #1195 to make one great PR?
> 
> 
> 
> I've submitted a PR to this branch which merges the changes and keep broadcast as it was.

Great, thanks for that!

File: tinygrad/lazy.py [hunk 3]
```diff
@@ -192,6 +192,7 @@ def toCPU(self):
   def cast(self:LazyBuffer, arg:DType) -> LazyBuffer: return elementwise_op(UnaryOps.CAST, self, arg=arg) if self.dtype != arg else self
   def unary_op(self:LazyBuffer, op:UnaryOps) -> LazyBuffer: return elementwise_op(op, self)
   def binary_op(self:LazyBuffer, op:BinaryOps, y:LazyBuffer) -> LazyBuffer: return elementwise_op(op, self, y)
+  def ternary_op(self:LazyBuffer, op:TernaryOps, y: LazyBuffer, z:LazyBuffer) -> LazyBuffer: return create_lazybuffer(self.device, ShapeTracker(self.shape), TernaryOps, LazyOp(op, (self, y, z,), None), y.dtype)
```
Comment (akretz) [inline tinygrad/lazy.py:195]: Shouldn't this also call `elementwise_op` to allow for merged kernels?

**Comment by akretz** [comment]
@flammit I've rebased to the current master, solved the conflicts and added the op to the new WebGPU backend.

File: test/test_ops.py [hunk 4]
```diff
@@ -97,7 +97,7 @@ def test_where(self):
       lambda x: torch.where(x > 0.5, 4, 2),
       lambda x: (x > 0.5).where(4, 2), forward_only=True)
 
-    for shps in [[(10,),(1,),(1,)], [(10,10),(10,),(10,)], [(100,)]*3, [(10,10)]*3]:
+    for shps in [[(8,),(1,),(1,)], [(10,10),(10,),(10,)], [(100,)]*3, [(10,10)]*3]:
```
Comment (wozeparrot) [inline test/test_ops.py:100]: What's the reason behind this shape change?

Comment (flammit) [reply test/test_ops.py:100]: To ensure that there's a float4 test case which covers the linearizer.py:455 change.

File: tinygrad/lazy.py (hunk 3 above)
Comment (geohot) [reply tinygrad/lazy.py:195]: Oh yea, it has to merge.

**Comment by geohot** [comment]
Getting close. But ternary ops need to merge with other elementwise ops.

File: tinygrad/tensor.py (hunk 2 above)
Comment (Antonio12389) [reply tinygrad/tensor.py]: Un favor me puedes ayudar veo que no puedo recuperar mis cuentas de criptómonedas me pueden apoyar y recuperar todo en si

**Comment by geohot** [comment]
Congrats, thanks!

**Pull request merged**